### PR TITLE
[fix][client] Add classLoader field for `SchemaDefinition`

### LINKED ISSUE #15899: SchemaDefinition.<T>builder().withJsonDef() does not register logical type conversions (e.g. java.lang.Instant)
**Describe the bug**
When using either `SchemaDefinition.<T>builder().withPojo()` (or directly `Schema.AVRO()`) time conversions for logical types are registered so POJOs with some java.time objects (not all) are correctly serialized.

However when method  `SchemaDefinition.<T>builder().withJsonDef()` is used instead such registration is not performed. We have to manually register those types for these conversions to work:

```
import org.apache.pulsar.shade.org.apache.avro.data.TimeConversions;
import org.apache.pulsar.shade.org.apache.avro.generic.GenericData;
import org.apache.pulsar.shade.org.apache.avro.reflect.ReflectData;


  private static final GenericData[] GD_INSTANCES = { ReflectData.AllowNull.get(), ReflectData.get(), GenericData.get() };

  // Static registration using shaded objects before reading any schema
  static {
    for (GenericData gd : GD_INSTANCES) {
      gd.addLogicalTypeConversion(new TimeConversions.DateConversion());
      gd.addLogicalTypeConversion(new TimeConversions.TimeMillisConversion());
      gd.addLogicalTypeConversion(new TimeConversions.TimestampMillisConversion());
    }
  }
```

Weirdly enough we have found that using a POJO with a `java.time.Instant` and using an schema with a LogicalType `timestamp-millis` it is converted to `Long` and assigned to field in POJO that has type `java.lang.Instant` (How can this even be possible?). 

**To Reproduce**
Create a POJO with an Instant field.

Produce to a topic using that POJO (and registering the Schema).

On the consumer side use the AVRO schema text with `SchemaDefinition.<T>builder().withPojo()`  to read from the topic. Review what is contained in the read objetcts.

For example. If POJO type is `TestPojo` we could use:

```
Schema<TestPojo> schema = AvroSchema.of(SchemaDefinition.<TestPojo>builder().withJsonDef(org.apache.avro.reflect.ReflectData.AllowNull.get().getSchema(TestPojo.class)).build());
```

And use that `Schema` to create a consumer.

**Expected behavior**
That same logical conversions are applied both using `withJsonDef()` or with `withPojo()`

**Screenshots**
Not applicable

**Desktop (please complete the following information):**
Windows

**Additional context**
This related to #15858

### PR DESCRIPTION
**Pull request by coderzc**
Fixes #15899

### Motivation

Now, don‘t register logical type conversions when use `SchemaDefinition.<T>builder().withJsonDef()` beacase it without  classLoader param.

See:
https://github.com/apache/pulsar/blob/04aa9e8e51869d1621a7e25402a656084eebfc09/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/reader/AvroReader.java#L58-L68

We can add the classLoader field for `SchemaDefinition`, user can manually pass a classLoader to register logical type conversions

### Modifications

Add classLoader field for `SchemaDefinition`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change added tests and can be verified

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

**Review by shibd — APPROVED** [review]
/LGTM

**Comment by coderzc** [comment]
> PR

What means?
This PR modifies `SchemaUtil` rather than `StructSchema`.

**Review by mattisonchao — APPROVED** [review]

**Comment by mattisonchao** [comment]
Hi @coderzc 

We re-run the test many times, but the test still fails. Could you take a look?

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/SchemaDefinitionBuilder.java [hunk 1]
```diff
@@ -80,6 +80,15 @@
...
     SchemaDefinitionBuilder<T> withPojo(Class pojo);
 
+    /**
+     * Set schema of pojo classLoader.
+     *
+     * @param classLoader pojo classLoader
+     *
+     * @return schema definition builder
+     */
+    SchemaDefinitionBuilder<T> withClassLoader(ClassLoader classLoader);
```
Comment (codelipenghui) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/SchemaDefinitionBuilder.java:90]: If need to add a public API to fix the issue, it should start with a proposal first.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

**Review by codelipenghui — APPROVED** [review]

**Comment by coderzc** [comment]
@congbobo184 @mattisonchao PTAL

**Pull request merged**